File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/math_on_list.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception
import kotlin.math.pow
import kotlin.math.sqrt

suspend fun Piper.executeMathOnList(statement: JSONObject, onError: (e: Exception) -> Unit): Any? {
    try {
        Log.d("hello", "execute_math_on_list")
        val pointer = statement.getJSONObject("pointer")
        val op = pointer["op"].toString()
        val values = executeStatement(pointer["list"].toString()) as List<Any?>

        val numberList = ArrayList<Number>()

        for (v in values) {
            when (v) {
                is Double -> numberList.add(v)
                is Int -> numberList.add(v)
                else -> return null
            }
        }

        val out = when (op) {
            "SUM" -> numberList.sumOf { it.toDouble() }
            "MIN" -> numberList.minOf { it.toDouble() }
            "MAX" -> numberList.maxOf { it.toDouble() }
            "AVERAGE" -> numberList.sumOf { it.toDouble() } / numberList.size
            "MEDIAN" -> {
                val sortedList = numberList.sortedBy { it.toDouble() }
                if (sortedList.size % 2 == 0) {
                    return (sortedList[sortedList.size / 2].toDouble() + sortedList[sortedList.size / 2 - 1].toDouble()) / 2
                } else {
                    return sortedList[sortedList.size / 2]
                }
            }
            "MODE" -> {
                val sortedList = numberList.sortedBy { it.toDouble() }
                var count = 1
                var mode = sortedList[0]
                for (i in 1 until sortedList.size) {
                    if (sortedList[i] == sortedList[i - 1]) {
                        count++
                    } else {
                        count = 1
                    }

                    if (count > 1) {
                        mode = sortedList[i]
                        break
                    }
                }
                return mode
            }
            "STD_DEV" -> {
                val mean = numberList.sumOf { it.toDouble() } / numberList.size
                val meanList = ArrayList<Double>()
                for (x in numberList) {
                    meanList.add((x.toDouble() - mean).pow(2))
                }
                return sqrt(meanList.sum() / numberList.size)
            }
            "RANDOM" -> numberList.random()
            else -> null
        } ?: return null

        return if (out.toString().endsWith(".0")) {
            out.toInt()
        } else {
            out.toDouble()
        }
    } catch (e: Exception) {
        onError(e)
        return null
    }
}

File: piper/extensions/chrome/src/engine/statements/math_on_list.js
import { stringify } from "../util"


function mathModes(values) {
  var modes = [];
  var counts = [];
  var maxCount = 0;
  for (var i = 0; i < values.length; i++) {
    var value = values[i];
    var found = false;
    var thisCount;
    for (var j = 0; j < counts.length; j++) {
      if (counts[j][0] === value) {
        thisCount = ++counts[j][1];
        found = true;
        break;
      }
    }
    if (!found) {
      counts.push([value, 1]);
      thisCount = 1;
    }
    maxCount = Math.max(thisCount, maxCount);
  }
  for (var j = 0; j < counts.length; j++) {
    if (counts[j][1] === maxCount) {
        modes.push(counts[j][0]);
    }
  }
  return modes;
}


function mathStandardDeviation(numbers) {
    var n = numbers.length;
    if (!n) return null;
    var mean = numbers.reduce(function(x, y) {return x + y;}) / n;
    var variance = 0;
    for (var j = 0; j < n; j++) {
      variance += Math.pow(numbers[j] - mean, 2);
    }
    variance = variance / n;
    return Math.sqrt(variance);
}


export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let op = stringify(pointer.op)
        let values = await this.executeStatement(stringify(pointer.list))

        switch(op) {
            case "SUM":
                return values.reduce((a, b) => a + b, 0)
            case "MIN":
                return Math.min(...values)
            case "MAX":
                return Math.max(...values)
            case "AVERAGE":
                return values.reduce((a, b) => a + b, 0) / values.length
            case "MEDIAN":
                values.sort((a, b) => a - b)
                let half = Math.floor(values.length / 2)
                if (values.length % 2) {
                    return values[half]
                }
                return (values[half - 1] + values[half]) / 2.0
            case "MODE":
                return mathModes(values)
            case "STD_DEV":
                return mathStandardDeviation(values)
            case "RANDOM":
                return values[Math.floor(Math.random() * values.length)]
        }
    } catch (e) {
        onError(e)
    }

    return null
}
